f3d 1.0.1 (new formula)

Meakk opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

chenrui333 commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  test do
+    system "#{bin}/f3d", "--version"
```
Is there any better than just test the version?

SMillerDev replied:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Meakk replied:
yes I will improve that. There is currently no 3D file format in the test fixtures. I can manually create one in the formula but maybe it makes sense to add one for other software? (in OBJ format for example which is quite standard)

Meakk replied:
I have created a very basic OBJ file in the test.

carlocab commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,39 @@
+    inreplace "src/F3DOptions.cxx", "/etc/f3d/", "/usr/local/etc/f3d/"
+
+    args = std_cmake_args + %W[
+      -DCMAKE_BUILD_TYPE:STRING=Release
```
Specifying a `Release` build is already part of `std_cmake_args`.

Meakk replied:
thanks!

SMillerDev commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def install
+    # Is fixed upstream, remove this line for v1.1
+    inreplace "src/F3DOptions.cxx", "/etc/f3d/", "/usr/local/etc/f3d/"
```
```suggestion
    inreplace "src/F3DOptions.cxx", "/etc/f3d/", "#{etc}/f3d/"
```

Meakk replied:
thanks, it is indeed better!

SMillerDev commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,39 @@
+    end
+
+    mkdir_p prefix/"f3d"
+    mv prefix/"config.json", prefix/"f3d/config.json"
+    etc.install_symlink prefix/"f3d"
```
If this should be user configurable it shouldn't be symlinked because an update would override it then.

carlocab commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    mkdir_p etc/"f3d"
+    mv etc/"f3d/config.json", etc/"f3d/config.json.old" if File.exist?(etc/"f3d/config.json")
```
I'm confused by this. Is `config.json` supposed to be a user config?

If it is, then it is probably not a great experience to have your customised config be replaced by a template config every time there is a new version of f3d.

Meakk replied:
No, it is the system config. The user config is located in `~/.config/f3d/config.json` which is not installed and must be created manually.  
While we recommend using user config, it can make sense to modify the system config in order to have the same config for all users though.  
Is there a way to ask the user if he wants to replace the installed config by the default config during the installation?  
One solution could be to check if a system config already exists in which case we do not replace it.

carlocab replied:
Hmm. I see. I'm not sure this does what you want, though, as the variable `etc` resolves to `#{prefix}/Cellar/f3d/#{version}/etc`.

Meakk replied:
Thanks, there was an issue upstream that I have fixed using `inreplace`. There is no need for file copies now.

SMillerDev commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    mkdir_p etc/"f3d"
+    mv etc/"f3d/config.json", etc/"f3d/config.json.old" if File.exist?(etc/"f3d/config.json")
```
This also won't work because it'll only ever run on CI.

Meakk replied:
You mean that file system calls are not included when installing the binary package?

SMillerDev replied:
They are only ran on CI when creating the binary. You need a `post_install` block to achieve this. But if it's only system config it should be always overwritten instead.

Meakk replied:
Thanks, there was an issue upstream that I have fixed using `inreplace`. There is no need for file copies now.

Meakk commented:
@SMillerDev @chenrui333 @carlocab Thanks again, I appreciate you help. I have addressed all the issues you have reported. Let me know if it looks good to you now.

carlocab commented:
Think you should be good after fixing the line length. Thanks for your contribution.

carlocab commented on Formula/f3d.rb:
```diff
@@ -0,0 +1,48 @@
+  def install
+    # Is fixed upstream, remove this line for v1.1
+    inreplace "src/F3DOptions.cxx",
+    "directoryPath = vtksys::SystemTools::GetProgramPath(programFilePath);",
+    "programFilePath = vtksys::SystemTools::GetRealPath(programFilePath);"\
+    "directoryPath = vtksys::SystemTools::GetProgramPath(programFilePath);"
```
This might just be my personal preference (so feel free to ignore) but it would be good if you could indent this, and add a space before the `\` in the second line.

A `+` may also be more appropriate (instead of the `\`), but I don't really speak Ruby, so I could also be wrong on this one. It just strikes me as conveying string concatenation more clearly.

carlocab replied:
Whoops. I was wrong about `+`.

Sorry!

Meakk replied:
no problem ;)  
Fixed!

chenrui333 commented:
@Meakk, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!
